pkg/volume/rbd/rbd.go

```diff
@@ -235,3 +239,41 @@ func (plugin *rbdPlugin) execCommand(command string, args []string) ([]byte, err
 	cmd := plugin.exe.Command(command, args...)
 	return cmd.CombinedOutput()
 }
+
+type rbdRecycler struct {
+	name    string
+	spec    *volume.Spec
+	config  volume.VolumeConfig
+	host    volume.VolumeHost
+	timeout int64
+	volume.MetricsNil
+}
+
+func (r *rbdPlugin) NewRecycler(spec *volume.Spec) (volume.Recycler, error) {
+	return r.newRecyclerFunc(spec, r.host, r.config)
+}
+
+func newRecycler(spec *volume.Spec, host volume.VolumeHost, volumeConfig volume.VolumeConfig) (volume.Recycler, error) {
+	if spec.PersistentVolume == nil || spec.PersistentVolume.Spec.RBD == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.rbd is nil")
+	}
+
+	return &rbdRecycler{
+		name:    spec.Name(),
+		spec:    spec,
+		host:    host,
+		config:  volumeConfig,
+		timeout: volume.CalculateTimeoutForVolume(volumeConfig.RecyclerMinimumTimeout, volumeConfig.RecyclerTimeoutIncrement, spec.PersistentVolume),
+	}, nil
+}
+
+func (r *rbdRecycler) GetPath() string {
+	return ""
+}
+
+func (r *rbdRecycler) Recycle() error {
+	pod := r.config.RecyclerPodTemplate
+	pod.GenerateName = "pv-recycler-rbd-"
+	pod.Spec.Volumes[0].VolumeSource = api.VolumeSource{RBD: r.spec.PersistentVolume.Spec.RBD}
```

Comment:
Using `pod.Spec.Volumes[0]` directly is quite dangerous, the array may be uninitialized.
The only place suggesting that there *must* be at least one item in `pod.Spec.Volumes` is here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L409

@markturansky, how are the admins supposed to know this when preparing the template? IMO we miss some documentation or a check in `AttemptToLoadRecycler()` (as separate PR, of course).